File: front-end/src/decorators/regular-expression.decorator.ts
import { AbstractControl, ValidatorFn } from '@angular/forms';

import { DefineValidationMetadata } from './validation-factory';

/** Specifies that a data field value in ASP.NET Dynamic Data must match the specified
 *  regular expression.
 *  @param pattern {RegExp} the pattern the property must adhere to.
 *  @throws pattern has not been been assigned a value. */
export function RegularExpression(pattern: RegExp): PropertyDecorator {
    return (target: Object, propertyKey: string) => {
        if (!pattern) {
            throw new Error(`@${RegularExpression.name} on ${target.constructor.name} did not define a pattern.`);
        }

        const validator = regularExpressionValidator(pattern);
        DefineValidationMetadata(RegularExpression.name, validator, target, propertyKey);
    };
}


function regularExpressionValidator(pattern: RegExp): ValidatorFn {
    return (control: AbstractControl): { [name: string]: any } => {
        if (!control) {
            throw new Error('Control is null.');
        }

        if (control.value && !pattern.test(control.value)) {
            return {
                'regularExpression': `Doesnt match pattern '${pattern}'`
            };
        } else {
            return null;
        }
    };
}

File: TypeScriptGenerator/Decorators/RegularExpressionDecorator.cs
using System;
using System.ComponentModel.DataAnnotations;

namespace lvl.TypescriptGenerator.Decorators
{
    /// <summary>
    ///     Specifies that a data field value in ASP.NET Dynamic Data must match the specified regular expression.
    /// </summary>
    public class RegularExpressionDecorator : TypeScriptType
    {
        private string Pattern { get; }

        public RegularExpressionDecorator(string decoratorPath, RegularExpressionAttribute regularExpressionAttribute)
        {
            if (regularExpressionAttribute == null)
            {
                throw new ArgumentNullException(nameof(regularExpressionAttribute));
            }

            ModulePath = decoratorPath ?? throw new ArgumentNullException(nameof(decoratorPath));
            Pattern = regularExpressionAttribute.Pattern;
            Name = "RegularExpression";
        }

        public override string ToTypeScript() => $"@RegularExpression(/{Pattern}/)";
    }
}
